File: src/main/java/core/scm/specialforms/LetRec.java
package core.scm.specialforms;

import core.environment.Environment;
import core.evaluator.Evaluator;
import core.exceptions.IllegalSyntaxException;
import core.scm.SCMThunk;

import java.util.List;
import java.util.stream.IntStream;

import static core.scm.SCMUnspecified.UNSPECIFIED;

/* Syntax:
 * (letrec <bindings> <body>)
 *
 * <bindings>: ((<variable1> <init1>) ...)
 */
/*
 * TODO One restriction on letrec is very important:
 * it must be possible to evaluate each <init> without assigning or referring to the value of any <variable>.
 * If this restriction is violated, then it is an error.
 * The restriction is necessary because Scheme passes arguments by value rather than by name.
 * In the most common uses of letrec, all the <init>s are lambda expressions and the restriction is satisfied automatically.
 */
public enum LetRec implements ISpecialForm {
  LETREC;

  @Override
  public Object eval(List<Object> expression, Environment env, Evaluator evaluator) {
    if (expression.size() < 3) {
      throw IllegalSyntaxException.of(toString(), expression);
    }
    List<List> bindings = (List<List>)expression.get(1);
    /* Bind variables to fresh locations holding undefined values */
    Environment localEnv = new Environment(env);
    bindings.forEach(b -> localEnv.put(b.get(0), UNSPECIFIED));
    /* Evaluate inits */
    bindings.forEach(b -> localEnv.put(b.get(0), evaluator.eval(b.get(1), localEnv)));
    /* Evaluate body */
    IntStream.range(2, expression.size() - 1).forEach(i -> evaluator.eval(expression.get(i), localEnv));
    return new SCMThunk(expression.get(expression.size() - 1), localEnv);
  }

  @Override
  public String toString() {
    return "letrec";
  }
}

File: src/main/java/core/scm/specialforms/LetRec.kt
package core.scm.specialforms

import core.environment.Environment
import core.Evaluator
import core.exceptions.IllegalSyntaxException
import core.scm.Thunk
import core.Writer

/**
 * Syntax:
 * (letrec <bindings> <body>)
 *
 * <bindings>: ((<variable1> <init1>) ...)
 */
/*
 * TODO One restriction on letrec is very important:
 * it must be possible to evaluate each <init> without assigning or referring to the value of any <variable>.
 * If this restriction is violated, then it is an error.
 * The restriction is necessary because Scheme passes arguments by value rather than by name.
 * In the most common uses of letrec, all the <init>s are lambda expressions and the restriction is satisfied automatically.
 */
object LetRec : SpecialForm("letrec") {

    override fun eval(form: List<Any?>, evaluator: Evaluator): Any {
        if (form.size < 3 || form[1] !is List<*>) {
            throw IllegalSyntaxException(toString(), Writer.write(form))
        }
        val localEvaluator = Evaluator(Environment(evaluator.env))
        val bindings = form[1] as List<*>
        /* Bind variables to fresh locations holding undefined values */
        bindings.forEach {
            if (it !is List<*>) throw IllegalSyntaxException(toString(), Writer.write(form))
            localEvaluator.env[it[0]] = Unit
        }
        /* Evaluate inits */
        bindings.forEach {
            with (it as List<*>) { localEvaluator.env.put(it[0], localEvaluator.eval(it[1])) }
        }
        /* Evaluate body */
        for (i in 2..form.size - 2) {
            localEvaluator.eval(form[i])
        }
        return Thunk(form[form.size - 1], localEvaluator.env)
    }
}
